README.md

```diff
@@ -522,72 +525,169 @@ To add a `saml:AuthnContextDeclRef`, define `settings.authn_context_decl_ref`.
 building the authrequest object.
 
+## Service Provider Metadata
+
+To form a trusted pair relationship with the IdP, the SP (you) need to provide metadata XML
+to the IdP for various good reasons. (Caching, certificate lookups, relaying party permissions, etc)
 
-## Signing
+The class `OneLogin::RubySaml::Metadata` takes care of this by reading the Settings and returning XML.  All you have to do is add a controller to return the data, then give this URL to the IdP administrator.
+
+The metadata will be polled by the IdP every few minutes, so updating your settings should propagate
+to the IdP settings.
 
-The Ruby Toolkit supports 2 different kinds of signature: Embeded and `GET` parameters
+```ruby
+class SamlController < ApplicationController
+  # ... the rest of your controller definitions ...
+  def metadata
+    settings = Account.get_saml_settings
+    meta = OneLogin::RubySaml::Metadata.new
+    render :xml => meta.generate(settings), :content_type => "application/samlmetadata+xml"
+  end
+end
+```
 
-In order to be able to sign, define the private key and the public cert of the service provider:
+You can add `ValidUntil` and `CacheDuration` to the SP Metadata XML using instead:
 
 ```ruby
-  settings.certificate = "CERTIFICATE TEXT WITH HEAD AND FOOT"
-  settings.private_key = "PRIVATE KEY TEXT WITH HEAD AND FOOT"
+  # Valid until => 2 days from now
+  # Cache duration = 604800s = 1 week
+  valid_until = Time.now + 172800
+  cache_duration = 604800
+  meta.generate(settings, false, valid_until, cache_duration)
 ```
 
-The settings related to sign are stored in the `security` attribute of the settings:
+## Signing and Encryption
+
+Ruby SAML supports supports the signing and encryption functionality:
+
+1. Signing your SP Metadata XML
+2. Signing your SP SAML messages
+3. Encrypting IdP Assertion messages, and decrypting them upon receipt (EncryptedAssertion)
+4. Verifying signatures on IdP Assertion messages
+
+In order to use functions 1-3 above, you must first define your SP public certificate and private key:
 
 ```ruby
-  settings.security[:authn_requests_signed]   = true     # Enable or not signature on AuthNRequest
-  settings.security[:logout_requests_signed]  = true     # Enable or not signature on Logout Request
-  settings.security[:logout_responses_signed] = true     # Enable or not signature on Logout Response
-  settings.security[:want_assertions_signed]  = true     # Enable or not the requirement of signed assertion
-  settings.security[:metadata_signed]         = true     # Enable or not signature on Metadata
+  settings.certificate = "CERTIFICATE TEXT WITH BEGIN/END HEADER AND FOOTER"
+  settings.private_key = "PRIVATE KEY TEXT WITH BEGIN/END HEADER AND FOOTER"
+```
+
+Note that the same certificate and private key are used for all SP encryption and signing-related functions.
```

Comment:
Note that the certificate/private key that gonna be involved on the sign/validate encrypt/decrypt process gonna be the same.
IdP will encrypt using SP certificate and then SP will decrypt using the SP private_key.
SP will sign using the SP private_key, and the IdP will validate the signature with the SP certificate.
So you see that the SP certificate must be registered on the IdP side, to be used to encrypt and validate signatures.

Reply:
I've edited it to:

> Note that the same certificate (and its associated private key) are used to perform
> all decryption and signing-related functions (1-4) above. Ruby SAML does not currently allow
> to specify different certificates for each function.

The key point here is the last sentence; we are simply saying that you can't decrypt with Cert A and sign with Cert B.